grpc: add support for Any type mapping to google.protobuf.Any

ilCollez opened the pull request:
# Add Any Type Support for gRPC

This PR implements support for the `Any` type in gRPC code generation, addressing issue #3807 and #2188. The `Any` type is now properly mapped to `google.protobuf.Any` in generated protobuf files and includes automatic conversion between Go's `any` type and protobuf's `*anypb.Any`.

## Changes Made

### Core Implementation
- **Removed validation restriction** that prevented Any type usage in gRPC endpoints (`expr/grpc_endpoint.go`)
- **Enhanced protobuf mapping** to convert `expr.AnyKind` to `google.protobuf.Any` in proto files and `*anypb.Any` in Go code (`grpc/codegen/protobuf.go`)
- **Implemented type transformation** with JSON-based conversion between Go `any` and `*anypb.Any` via `structpb.Value` (`grpc/codegen/protobuf_transform.go`)
- **Added automatic imports** for `google/protobuf/any.proto`, `anypb`, `structpb`, and `encoding/json` when Any type is detected

### Code Generation
- **Proto files**: Automatically include `google/protobuf/any.proto` import and map Any fields to `google.protobuf.Any`
- **Go files**: Generate conversion code that marshals Go `any` to JSON, wraps in `structpb.Value`, and stores in `*anypb.Any`
- **Client/Server**: Support Any type in payloads, results, and error types with proper import management

### Testing & Documentation
- **Updated tests** to expect Any type support instead of failures
- **Added comprehensive test cases** for protobuf mapping and transformation
- **Enhanced documentation** with usage examples and limitations
- **Created integration example** demonstrating protobuf message transport via Any type

## Usage Example

```go
Method("processAnyData", func() {
    Payload(func() {
        Field(1, "data", Any, "Any type of data")
        Field(2, "metadata", MapOf(String, Any), "Metadata map")
    })
    Result(func() {
        Field(1, "processed", Any, "Processed data")
        Field(2, "success", Boolean, "Processing status")
    })
})
```

### Generated Proto

```proto
import "google/protobuf/any.proto";

message ProcessAnyDataRequest {
    optional google.protobuf.Any data = 1;
    map<string, google.protobuf.Any> metadata = 2;
}
```

### Generated Go Conversion Code

```go
// Go any → *anypb.Any
Data: func() *anypb.Any {
    if payload.Data == nil { return nil }
    if jsonData, err := json.Marshal(payload.Data); err == nil {
        if data, err := anypb.New(&structpb.Value{
            Kind: &structpb.Value_StringValue{StringValue: string(jsonData)},
        }); err == nil {
            return data
        }
    }
    return nil
}()
```

## Breaking Changes
None. This is a pure addition that enables previously unsupported functionality.

## Limitations
- JSON-based conversion means complex Go types may not roundtrip perfectly (numeric types become float64, custom structs become map[string]interface{})
- Best suited for JSON-compatible data like configuration, metadata, or loosely-typed content
## Testing
- All existing tests pass (no regressions)
- New comprehensive test coverage for Any type functionality
- Integration example demonstrates real-world usage

Commit:
grpc: add support for Any type mapping to google.protobuf.Any

- Remove validation restriction preventing Any type usage in gRPC endpoints
- Map expr.AnyKind to google.protobuf.Any in protobuf code generation
- Implement JSON-based conversion between Go any and *anypb.Any via structpb
- Add automatic imports for google/protobuf/any.proto and required Go packages
- Update tests to expect Any type support instead of failures
- Add comprehensive test cases for protobuf mapping and transformation
- Enhance documentation with usage examples and limitations

Fixes #3807

raphael commented:
This is looking good! Please run `make` locally and fix all issues before we can merge, thank you!

Commit:
grpc/codegen: fix unparam lint warning for newVar parameter

Add test coverage for both newVar=true and newVar=false scenarios
in protoBufTransform function to resolve golangci-lint unparam warning.

ilCollez commented:
Hi! I've fixed the issue with `golangci-lint`. Instead of removing the `newVar` parameter from `protoBufTransform` (since it's always `true`), I've preferred to add test cases where `newVar` is `false`. In this way I suppress the warning from `golangci-lint` and prevent changes to the function's signature. Is it okay for you?

raphael committed:
Merge branch 'v3' into feature/grpc-any-type-support
